### ISSUE TITLE
Excel export

### ISSUE THREAD

**Issue by stumpyfr** [strictdoc-project/strictdoc#258]
Hi,

Nice project, matching a clear need!
I would be interested to quickly add the excel export, it's a must-have for my current project.

Few questions before starting the implem:

1. the export command export everything, do you plan to make that more parametrizable?
2. adding a new format doesn't look very "plug&play", do you plan to move to a more dynamic system? (https://github.com/strictdoc-project/strictdoc/blob/master/strictdoc/core/actions/export_action.py#L64)
  2.1 happy to propose something if needed, would imagine a lookup in the `export/` to list the option, sort of related to question 1.
 3. I would imagine to only export one row per req. with an explicit ID, aligned with your idea?
  3.1 ID, title, statement, body (links? need to check if we can easily connect them in excel)
  
Best,

**Comment by stanislaw** [comment]
Hi @stumpyfr,

Thank you for opening your issue. I am excited to see someone evaluating StrictDoc!

Excel support was something that I was going to implement in the very near future and your questions are the same questions I was asking myself.

I have pushed some changes and your questions 1 and 2 should be answered by them (already in master).

- There is now `--formats` parameter which allows you to specify the format you want to export to.
  - html is the default, `--formats=rst` can be used for RST
- The lower-level logic has gone from the `ExportAction` class to the dedicated classes: HTML and RST generator.

On top of those changes, I have pushed this WIP branch: `https://github.com/strictdoc-project/strictdoc/compare/excel-generator?expand=1` where there is a basic Excel export skeleton.

A few comments:

- I have been using the `xlsxwriter` library for other projects and I would use it in StrictDoc as well. Let me know if you know any better alternatives.

- I am thinking of making the row 0 to list the headers: `UID, Title, Statement`. Let me know if this is something that makes sense for your use case as well.

- Could you share more details about your use case? 
  
  - Do you have parent links?
  - Can your requirements have multiple comments or just one?
  - Do you ever use `BODY:` or you only use single or multiline `STATEMENT`s?

If this branch is something that goes to the right direction, I welcome you to pick it up and push your own changes that you would like to see StrictDoc doing. Otherwise, I can complete the implementation myself, just let me know what is missing.

P.S. Aside from these technical questions, I would be interested in knowing more about your use case. What kind of requirements/specifications are you trying to model with StrictDoc?

**Comment by stumpyfr** [comment]
Pushed #267 , tell me if what you had in mind.

- `xlsxwriter` is totally fine
- I made a more configurable system, let's see how you would like to expose this configuration by the cli
  - for my need, UID and statement + parent should be fine, the rest -> "please open the HTML...and the docx soon as we expert to docx"
 - yes for the links, most of the time only one, luckily because excel doesn't allow more than one link :/
 - good question, I would say probably a lot of comment but would keep them out of excel, excel if really just the mapping table to be shared with the different actors of the project, HTML/Docx should be the owner of the "discussions"
 - statement is fine for me, the title is repetitive and the body is useful, not into excel (same logic as previously)
 
 
few feedbacks:

1. Is that possible to add "custom" field?
  1.1 I would need at least `Owner` and `Priority`  
2. the order of the field into the `sdoc` looks important, if you don't set each field in a very specific order, the DSL is failing, not ideal
3. the next big step for me would be a docx export, clearly a pain in the ass but...still useful.
 3.1 I was thinking about trying https://docxtpl.readthedocs.io/en/latest/, need to admit, no experience with

**Comment by stanislaw** [comment]
> Pushed #267 , tell me if what you had in mind.
> 
>     * `xlsxwriter` is totally fine

Good. I have left my comments there.

>     * I made a more configurable system, let's see how you would like to expose this configuration by the cli
>       
>       * for my need, UID and statement + parent should be fine, the rest -> "please open the HTML...and the docx soon as we expert to docx"

Ok for now

>     * yes for the links, most of the time only one, luckily because excel doesn't allow more than one link :/
> 
>     * good question, I would say probably a lot of comment but would keep them out of excel, excel if really just the mapping table to be shared with the different actors of the project, HTML/Docx should be the owner of the "discussions"

Ok for now. Eventually it will be made more configurable. Two known solutions for 'multiple' problem: either merging the cells or duplicating the row and only adding the second comment to the comment row. Both are hacks, but I have seen both to work ok (my current preference is the second one).

General note: a lot of things are missing. To keep the discussions shorter, please open separate tickets. All feedback is appreciated.
 
>     * statement is fine for me, the title is repetitive and the body is useful, not into excel (same logic as previously)

Ok for now.

> few feedbacks:
> 
>     1. Is that possible to add "custom" field?
>        1.1 I would need at least `Owner` and `Priority`

Please open a separate ticket. I have been thinking about this for quite a while and don't know an easy solution that will not result in a messy grammar. I have my own very custom fields, such as `ECSS_VERIFICATION_METHOD` which I would like to keep out of the default StrictDoc's grammar.

>     2. the order of the field into the `sdoc` looks important, if you don't set each field in a very specific order, the DSL is failing, not ideal

This was my intention to keep the default grammar very strict including the order. I would like to keep it like this for the default grammar but probably relax it for the custom fields.

>     3. the next big step for me would be a docx export, clearly a pain in the ass but...still useful.
>        3.1 I was thinking about trying https://docxtpl.readthedocs.io/en/latest/, need to admit, no experience with

Another thing on my TODO list as well. Please go ahead and suggest something. My only requirement here would be to allow generation of company-branded Word templates but this is a Nice To Have, not a MUST requirement.

**Comment by stanislaw** [comment]
@stumpyfr I have added a basic integration test on top of your branch and have merged it already.

Please test the master branch and close this ticket if everything is fine.

If there is any new work, please add new tickets.

**Comment by stumpyfr** [comment]
LGTM
